src/com/amazon/corretto/crypto/provider/HkdfSpec.java

```diff
@@ -0,0 +1,152 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.corretto.crypto.provider;
+
+import java.security.spec.KeySpec;
+
+public class HkdfSpec implements KeySpec {
+  final int mode;
+  final byte[] secretOrPrk;
+  final byte[] salt;
+  final byte[] info;
+  final int desiredLength;
+  final String algorithmName;
+  public static final String DEFAULT_ALGORITHM_NAME = "Hkdf";
+  public static final int HKDF_MODE = 1;
+  public static final int HKDF_EXTRACT_MODE = 2;
+  public static final int HKDF_EXPAND_MODE = 3;
+
+  public HkdfSpec(
+      final int mode,
+      final byte[] secret,
+      final byte[] salt,
+      final byte[] info,
+      final byte[] prk,
+      final int desiredLength,
+      final String algorithmName) {
+
+    switch (mode) {
+      case HKDF_MODE:
+        this.secretOrPrk = Utils.requireNonNull(secret, "secret cannot be null for HKDF");
+        this.salt = Utils.requireNonNull(salt, "salt cannot be null for HKDF");
+        this.info = Utils.requireNonNull(info, "info cannot be null for HKDF");
+        this.desiredLength = validateDesiredLength(desiredLength);
+        break;
+      case HKDF_EXTRACT_MODE:
+        this.secretOrPrk = Utils.requireNonNull(secret, "secret cannot be null for HKDF_EXTRACT");
+        this.salt = Utils.requireNonNull(salt, "salt cannot be null for HKDF_EXTRACT");
+        this.info = null;
+        this.desiredLength = 0;
+        break;
+      case HKDF_EXPAND_MODE:
+        this.secretOrPrk = Utils.requireNonNull(prk, "prk cannot be null for HKDF_EXPAND");
+        this.salt = null;
+        this.info = Utils.requireNonNull(info, "info cannot be null for HKDF_EXPAND");
+        this.desiredLength = validateDesiredLength(desiredLength);
+        break;
+      default:
+        throw new IllegalArgumentException("mode is not a valid value");
+    }
+    this.mode = mode;
+    this.algorithmName = algorithmName != null ? algorithmName : DEFAULT_ALGORITHM_NAME;
+  }
+
+  private static int validateDesiredLength(final int desiredLength) {
```

Comment:
Still missing the upper bound check. Otherwise review lgtm

Reply:
To do the upper bound check, the size of the digest is needed. The HkdfSpec class doesn't have the digest size. The [HkdfSecretKeyFactorySpi](https://github.com/amirhosv/amazon-corretto-crypto-provider/blob/hkdf/src/com/amazon/corretto/crypto/provider/HkdfSecretKeyFactorySpi.java#L84-L91) class has that information and that's where the check happens. The `checkExpandLength` is invoked when HKDF and HKDF_Expand.